### ISSUE TITLE
Add GPU support

### ISSUE THREAD

**Issue by fepegar** [fepegar/torchio#388]
**🚀 Feature**

Add support to transform images in GPU devices.

**Motivation**

Some transforms would benefit a lot from hardware acceleration. I suspect the two main sources of improvement would be resampling (`Resample`, `RandomElasticDeformation`, `RandomAffine`, `RandomAnisotropy`, `RandomMotion`) and Fourier transforms (`RandomMotion`, `RandomSpike`, `RandomGhosting`).

Most users want to keep their GPU for training and not for preprocessing / augmentation, but if there are enough resources available, it's nice to add GPU support to transforms.

**Pitch**

Supporting FFT is easy and seems to help, using the new `torch.fft` module in PyTorch 1.7. I have added some support in the `fourier` branch.

On CPU:

```python
In [1]: import torchio as tio
   ...: t1 = tio.datasets.FPG().t1
   ...: t1.load()
   ...: transform = tio.RandomSpike()
   ...: %timeit transform(t1)
1.35 s ± 3.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

On GPU:
```python
In [1]: import torchio as tio
   ...: t1 = tio.datasets.FPG().t1
   ...: t1.load()
   ...: transform = tio.RandomSpike()
   ...: %timeit transform(t1)
155 ms ± 820 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Supporting resampling is more complex, because of the way PyTorch deals with coordinates etc. Ideally, we would be able to convert from a "world" affine transform to a PyTorch one.

Some discussions about converting to/from PyTorch conventions for affine transformations:
- [Affine transformation matrix paramters conversion](https://discuss.pytorch.org/t/affine-transformation-matrix-paramters-conversion/19522/18)
- [Unexpected behaviour for affine_grid and grid_sample with 3D inputs](https://github.com/pytorch/pytorch/issues/35775#issuecomment-705702703)
- [Generating pytorch's theta from affine transform matrix](https://stackoverflow.com/questions/61964914/generating-pytorchs-theta-from-affine-transform-matrix)

The steps for this transition to happen would be:
1. Make sure everything works normally with tensors on GPU
2. Make sure the run time for FFT transforms is improved
3. Figure out how to resample medical images properly using PyTorch
4. Make sure the run time for resampling transforms is improved
5. Check if run time for other transforms gets better as well
6. Test that everything works as before, on multiple PyTorch versions

**Comment by romainVala** [comment]
Great ! this will be a nice improvement

The thing with gpu support, is that it will not be available in pytorch DataLoader since they are cpu only . (correct ?)

About your example on fft, the code is exactly the same for cpu and gpu ... strange ... don't you need to explicitly say you are in gpu / cpu  ?

I played a little bit with affine_grid, and It is quite difficult to correctly understand the convention. I could not make it exactly the same (half voxel difference) but from the thread you pointed I am sure you will get the correct conversion ... good luck

**Comment by fepegar** [comment]
> The thing with gpu support, is that it will not be available in pytorch DataLoader since they are cpu only . (correct ?)

Oh yeah I've heard that. If that's actually the case, maye there is no point in doing this.

> About your example on fft, the code is exactly the same for cpu and gpu ... strange ... don't you need to explicitly say you are in gpu / cpu ?

No, the PyTorch API typically doesn't mind about the tensors device.

> I played a little bit with affine_grid, and It is quite difficult to correctly understand the convention. I could not make it exactly the same (half voxel difference) but from the thread you pointed I am sure you will get the correct conversion ... good luck

I'm not planning to work on this immediately... If anyone is interested, please say so!

**Comment by romainVala** [comment]
> maye there is no point in doing this.

no, I am very confident that this will be useful. may be the DataLoader strategy will change in a near future, but even without it, there are case where it make sens to have gpu transform
For instance for the reverse problem (#299). Since you just evaluate your model with GPU and you want to spatially inverse the prediction, you will be happier with a GPU transform (for sure)

I have no time immediately too, but I hope in a near future ... once the conversion between nifti affine and pytorch resamping is done, it may not be too difficult to program it ...

Just out of curiosity, do you know if there are tools in pytorch to perform non linear deformation. (did you see example of non linear deformation of the grid ?)

**Comment by fepegar** [comment]
> For instance for the reverse problem (#299). Since you just evaluate your model with GPU and you want to spatially inverse the prediction, you will be happier with a GPU transform (for sure)

Yeah, I suppose inference would be faster. But I suppose inference is not a bottleneck, I'd more interested in making training faster.

> Just out of curiosity, do you know if there are tools in pytorch to perform non linear deformation. (did you see example of non linear deformation of the grid ?)

I think it's possible. Look at Table 1 of [Spatial Transformer Networks](https://arxiv.org/pdf/1506.02025.pdf). In the corresponding [PyTorch tutorial](https://pytorch.org/tutorials/intermediate/spatial_transformer_tutorial.html) they seem to use affine only, though.

You could ask someone from DeepReg or Kornia, they will probably know this stuff.

**Comment by efirdc** [comment]
I think a good reason for doing this would be to support differentiable augmentation. This was done in [StyleGAN2-ADA](https://github.com/NVlabs/stylegan2-ada) to train GANs using limited data. The augmentations are performed on the generated images before they are input to the discriminator, so they have to be differentiable.

**Comment by mfazampour** [comment]
I second @efirdc . it would be a really useful feature to be able to apply the transforms without the need to detach.

**Comment by justusschock** [comment]
> The thing with gpu support, is that it will not be available in pytorch DataLoader since they are cpu only . (correct ?)

This is only partially correct. The DataLoader has the limitation, that if it performs GPU-Ops, every worker process there creates its own CUDA-Context (afaik). So the only way to overcome this is to apply those GPU transforms within the main process (which also plays nicer with pinned memory). I have implemented this already in two different frameworks ([[1]](https://github.com/PhoenixDL/rising/blob/master/rising/loading/loader.py#L51), [[2]](https://github.com/PyTorchLightning/lightning-flash/blob/d0e6ee65419e140ea7894f1c03a5955919d66310/flash/core/data/process.py#L260)). So this is definitely possible, but either requires a custom collate fn (bad user experience) or subclassing the loader if you are not controlling the training flow (like [2] does).

> Just out of curiosity, do you know if there are tools in pytorch to perform non linear deformation. (did you see example of non linear deformation of the grid ?)

This is possible with torch.nn.functional.grid_sample where you can specify an arbitrary deformation field.

> Yeah, I suppose inference would be faster. But I suppose inference is not a bottleneck, I'd more interested in making training faster.
I imagine this only being faster if it is on the same GPU. With medical images, this might be challenging, since you run OOM pretty fast.

**Comment by romainVala** [comment]
ok, If I understand correctly, you say, one should put num_worker=0, to stay in the main process, ...
that make sens, especially if one have only one GPU, to avoid multi threading 

Did you test it with torchio transform ?  

from what I understand, this can be easy for some transform but not for other, like resampling, or elastic deformation that heavily rely either on nibable, or sitk, ... but there is no sitk version for gpu ... 

> This is possible with torch.nn.functional.grid_sample where you can specify an arbitrary deformation field.

Yes I know this function, and I tried once to use it for affine resampling, but it is tricky to get the same convention as the one used in torchio (ie sitk). I think this is the only option to go with resampling in gpu (rigth ?)
so a first start would be to make an affine resampling with pytorch grid_sample ..

And for non linear, if you know some code that implement a similare thing as in ElasticDeformation, I would be interested to look at it (or same for just the affine part ...)

would you have time, to make change in torchio for gpu support ?

**Comment by justusschock** [comment]
> ok, If I understand correctly, you say, one should put num_worker=0, to stay in the main process, ...

Either that or use the transforms after the ones from the dataloader. that being said, GPU transforms are also more efficient on batches, which torchio does not (yet) approach at all.

> Did you test it with torchio transform ?
Only with some custom ones.

> from what I understand, this can be easy for some transform but not for other, like resampling, or elastic deformation that heavily rely either on nibable, or sitk, ... but there is no sitk version for gpu .

Exactly. All transforms must be either torch or jax-like numpy (preferably torch) for GPUs. So we have to rewrite it completely.

> And for non linear, if you know some code that implement a similare thing as in ElasticDeformation, I would be interested to look at it (or same for just the affine part ...)

There is [airlabs](https://github.com/airlab-unibas/airlab) which implements some registration. I also have my own implementation of that, but unfortunately, I cannot share it just yet.

> would you have time, to make change in torchio for gpu support ?

Probably not. This goes far beyond just enabling a few changes and includes rewriting all transforms that currently rely on nibabel or sitk.

**Comment by fepegar** [comment]
If we ever start tackling this, here are a couple of sources that might be useful:
1. [PyTorch forum – Affine transformation matrix parameters conversion](https://discuss.pytorch.org/t/affine-transformation-matrix-paramters-conversion/19522/18?u=fepegar)
1. [StackOverflow – Generating pytorch's theta from affine transform matrix](https://stackoverflow.com/a/61973182/3956024)

**Comment by jizongFox** [comment]
Hello all, I am wondering if it would be possible to just replace numpy by cupy to enjoy the gpu acceleration without major change of code? It seems they share similar API and I am not sure if cupy can meet multiprocessing problems.

**Comment by justusschock** [comment]
Hi @jizongFox ,

I’m afraid this won’t be possible/sufficient. Torchio heavily uses nibabel and SimpleITK which would still have a reference to normal bumpy. 
But more importantly both of them have a lot of custom C++ code that would still be executed on cpu (in fact, SimpleITK is almost only a python wrapper for itk’s c++ backend

**Comment by fepegar** [comment]
Closing for now as it would take a considerable effort.
